ChattoAdditions/Source/Input/Photos/PhotosInputWithPlaceholdersDataProvider.swift

```diff
@@ -0,0 +1,118 @@
+/*
+ The MIT License (MIT)
+
+ Copyright (c) 2015-present Badoo Trading Limited.
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+import PhotosUI
+
+protocol PhotosInputDataProviderDelegate: class {
+    func handlePhotosInputDataProviderUpdate(_ dataProvider: PhotosInputDataProviderProtocol, updateBlock: @escaping () -> Void)
+}
+
+protocol PhotosInputDataProviderProtocol: class {
+    weak var delegate: PhotosInputDataProviderDelegate? { get set }
+    var count: Int { get }
+    @discardableResult
+    func requestPreviewImage(at index: Int,
+                             targetSize: CGSize,
+                             completion: @escaping PhotosInputDataProviderCompletion) -> PhotosInputDataProviderImageRequestProtocol
+    @discardableResult
+    func requestFullImage(at index: Int,
+                          progressHandler: PhotosInputDataProviderProgressHandler?,
+                          completion: @escaping PhotosInputDataProviderCompletion) -> PhotosInputDataProviderImageRequestProtocol
+    func fullImageRequest(at index: Int) -> PhotosInputDataProviderImageRequestProtocol?
+    func cancelImageRequest(_ request: PhotosInputDataProviderImageRequestProtocol)
+}
+
+typealias PhotosInputDataProviderProgressHandler = (Double) -> Void
+typealias PhotosInputDataProviderCompletion = (PhotosInputDataProviderResult) -> Void
+
+enum PhotosInputDataProviderResult {
+    case success(UIImage)
+    case error(Error?)
+
+    var image: UIImage? {
+        guard case let .success(resultImage) = self else { return nil }
+        return resultImage
+    }
+}
+
+protocol PhotosInputDataProviderImageRequestProtocol: class {
+    var isFullImageRequest: Bool { get }
```

Comment:
Can you replace this variable with `func cancel()` that will hide implementation details about requests cancellation? This variable is only used to distinguish preview and full photos requests cancellation.

Reply:
Great suggestion! Done.